kdf/kdf.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2019 Thales e-Security, Inc
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+package kdf
+
+import (
+	"encoding/binary"
+	"crypto"
+	"errors"
+	"hash"
+	"io"
+	// Force registration of SHA1 and SHA2 families of cryptographic primitives to
+	// reduce the burden on KDF consuming packages.
+	_ "crypto/sha1"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+)
+
+var (
+	// ErrInvalidLengthParameter the KDF length parameter is invalid
+	ErrInvalidLengthParameter = errors.New("invalid length parameter")
+
+	// ErrInvalidSeedParameter a parameter is invalid.
+	ErrInvalidSeedParameter = errors.New("invalid input parameter")
+	)
```

Comment:
Do we need to export these errors and commit to them being part of our API? I think we should un-export these, unless you can suggest some scenarios where someone's program might respond to the error.

Reply:
Removed errors from being exported